sql: match postgres for casting 0 to regtype/regproc/regnamespace/regtype

lpessoa opened the pull request:
Fixes #67715 
 
Postgres outputs `-` for calls to
- `select 0::oid::regproc;`
- `select 0::oid::regprocedure;`
- `select 0::oid::regnamespace;`
- `select 0::oid::regclass;`
- `select 0::oid::regtype;`

CockroachDB outputs `0` instead.

In order to be compliant to Postgres behavior we need to change Cockroach to provide the same output for the above mentioned
scenarios.

For this we changed Datum to Format `DInts` as `-` for the special `0` scenario.
Additional tests where added to `cast` testdata and also a new `FakeEventPlanner`struct class was implemented to support the new test scenarios thus avoiding nil pointer exceptions during tests.

Release note: None

rafiss reviewed (changes requested):
Thanks for your contribution! I'm not sure if changing the `Format` logic is what we'd want here, but I am open to hearing an argument why. Did you see my comment on https://github.com/cockroachdb/cockroach/issues/67715#issuecomment-931576240 ? this might work better as a change to `PerformCast` in casts.go

For example, Postgres has this behavior, and CRDB still should also:
```
> select 0::oid::text
+--------+
| text   |
|--------|
| 0      |
+--------+
```

can you add a test for that?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71873#-:-MmdHgP5935fWxMib6u4:bfrz5fv)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @lpessoa)

___
*[pkg/sql/sem/tree/eval.go, line 3618 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71873#-MmdI-kz1tbABbRrrUSH:-MmdI-kz1tbABbRrrUSI:b-ui4bmg) ([raw file](https://github.com/cockroachdb/cockroach/blob/7893b4d157c20584109e9117c3a18aa4f96e0829/pkg/sql/sem/tree/eval.go#L3618)):*
> ```Go
> 		Settings:         st,
> 		NodeID:           base.TestingIDContainer,
> 		Planner:          &FakeEvalPlanner{},
> ```

where is the Planner used in the new tests?

are you sure this wouldn't work:
```
Planner: &faketreeeval.DummyEvalPlanner{},
```


<!-- Sent from Reviewable.io -->

lpessoa committed:
sql: fix casting of zero OID

Release note (sql change): To align with PostgreSQL, casting an OID type
with a value of `0` to a regtype, regproc, regclass, or regnamespace now
will convert the value to the string `-`. The reverse behavior is
implemented too, so a `-` will become `0` if casted to a `reg` OID type.

lpessoa commented:
> Thanks for the changes! I'll give this a closer look soon. One more improvement we should make is to support the reverse behavior:
> 
> ```
> postgres=# select '-'::regclass::oid;
>  oid
> -----
>    0
> ```

@rafiss added the reverse behavior too.

rafiss reviewed (changes requested):
thanks! this is close, just a few minor comments

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71873#-:-MnRUp2c0WX001ZjeGpM:b-qqxyr5)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @lpessoa)

___
*[pkg/sql/sem/tree/cast.go, line 2350 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/71873#-MnRTzGiDi5MIfQWJ_jW:-MnRTzGiDi5MIfQWJ_jX:b-3fwhfd) ([raw file](https://github.com/cockroachdb/cockroach/blob/54b530084fcc7e654136092aadf8dcd9e7f1aa44/pkg/sql/sem/tree/cast.go#L2350)):*
> ```Go
> 			if string(*v) == ZeroDOidValue {
> 				// zero oid value is represented by '-' in postgres.
> 				return ParseDOid(ctx, "0", t)
> ```

i believe this can be replaced with
```
				tmpOid := NewDOid(0)
				tmpOid.semanticType = t
				return tmpOid, nil
```

it will avoid unneeded lookups
___
*[pkg/sql/sem/tree/cast.go, line 2426 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/71873#-MnRSfCvBWmq2HKtsNbT:-MnRSfCwEAOV7cZDv081:bbxnzgh) ([raw file](https://github.com/cockroachdb/cockroach/blob/54b530084fcc7e654136092aadf8dcd9e7f1aa44/pkg/sql/sem/tree/cast.go#L2426)):*
> ```Go
> 		// ctx.Planner will be nil only for 'make test PKG=./pkg/sql/sem/tree TESTS=TestEval'
> 		// we just perform a nil check and keep backward compatibility
> 		if ctx.Planner != nil {
> ```

i'd prefer not to add this condition

can you remove the nil-check, and instead put your tests in `pkg/sql/logictest/testdata/logic_test/cast`
___
*[pkg/sql/sem/tree/testdata/eval/cast, line 1401 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/71873#-MnRSlhh2CsTSGubqn2_:-MnRSlhh2CsTSGubqn2a:b-ucuham) ([raw file](https://github.com/cockroachdb/cockroach/blob/54b530084fcc7e654136092aadf8dcd9e7f1aa44/pkg/sql/sem/tree/testdata/eval/cast#L1401)):*
> ```
> ----
> '-'
> ```

can you also add a test for `select '-'::regclass::oid;`


<!-- Sent from Reviewable.io -->

lpessoa commented:
shoot... forgot to remove this one from the previous iteration... thanks for pointing it out
```go
if ctx.Planner != nil {
```

lpessoa commented:
@rafiss do you know the  reason why `pkg/sql/logictest/testdata/logic_test/cast` is encoded using **windows-1252**?
It seems to be the only one using that encoding. Some test will break if i change it to UTF-8.

rafiss commented:
> do you know the reason why pkg/sql/logictest/testdata/logic_test/cast is encoded using windows-1252 

The reason is because of this test: https://github.com/cockroachdb/cockroach/blob/ce0b24e0f2a16e65a3bd34ff835c1609ea518342/pkg/sql/logictest/testdata/logic_test/cast#L266-L276

It tests a weird edge case with single-byte characters

I think we can use a different encoding by changing the query to be:
```
SELECT i, i::"char"::bytea, length(i::"char") ...
```

rafiss reviewed (changes requested):
thanks! i think this is getting close

firstly, please squash all your changes into one commit. then just a few more comments

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71873#-:-MnbieMu18sa3J1Gux0G:b3snnt5)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @lpessoa)

___
*[pkg/sql/logictest/testdata/logic_test/cast, line 342 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/71873#-MnbimHn4yLIj6JHvem4:-MnbimHn4yLIj6JHvem5:bmm989u) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e2252a4a470cba43191d59700755d2bb20809cf/pkg/sql/logictest/testdata/logic_test/cast#L342)):*
> ```
> SELECT '-'::regclass::oid
> ----
> 0
> ```

all of these tests can be condensed a bit. you can just add these tests:

```
query IOTTTTTT
SELECT i, i::oid, i::oid::text,
  i::oid::regproc, i::oid::regprocedure, i::oid::regnamespace, i::oid::regclass, i::oid::regtype
FROM (VALUES (0), (1)) v(i)
----
0  0  0  -  -  -  -  -
1  1  1  1  1  1  1  1

query TOOOOO
SELECT i, i::regproc::oid, i::regprocedure::oid, i::regnamespace::oid, i::regclass::oid, i::regtype::oid
FROM (VALUES ('-')) v(i)
----
-  0  0  0  0  0

statement error invalid input syntax for type oid: "-"
SELECT '-'::oid
```
___
*[pkg/sql/sem/tree/cast.go, line 2348 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/71873#-MnbjGtg2exts8ZQFYod:-MnbjGth-wtHA_7urBsy:b-gkiwhr) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e2252a4a470cba43191d59700755d2bb20809cf/pkg/sql/sem/tree/cast.go#L2348)):*
> ```Go
> 			return performIntToOidCast(ctx, t, i)
> 		case *DString:
> 			if string(*v) == ZeroDOidValue {
> ```

i think we need one more condition here -- this cast should only work `if t.Oid() != oid.T_oid`

the reason is that we still want `SELECT '-'::oid` to result in an error
___
*[pkg/sql/sem/tree/cast.go, line 2349 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/71873#-Mnbjf0866lDVAe1_BDs:-Mnbjf0866lDVAe1_BDt:b21zdx5) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e2252a4a470cba43191d59700755d2bb20809cf/pkg/sql/sem/tree/cast.go#L2349)):*
> ```Go
> 		case *DString:
> 			if string(*v) == ZeroDOidValue {
> 				// zero oid value is represented by '-' in postgres.
> ```

nit: use correct grammar:
```
// The zero oid value is represented by '-' in order to match Postgres.
```
___
*[pkg/sql/sem/tree/datum.go, line 5044 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/71873#-MnbjnNd9FS2418i6-uI:-MnbjnNd9FS2418i6-uJ:b-qqx43e) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e2252a4a470cba43191d59700755d2bb20809cf/pkg/sql/sem/tree/datum.go#L5044)):*
> ```Go
> }
> 
> // ZeroDOidValue postgres represents 0 oid value as '-'
> ```

nit: replace with
```
// ZeroOidValue postgres represents the 0 oid value as '-'.
```
___
*[pkg/sql/sem/tree/datum.go, line 5070 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/71873#-Mnbjp421Q6-qV11M_Ex:-Mnbjp421Q6-qV11M_Ey:be4avy0) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e2252a4a470cba43191d59700755d2bb20809cf/pkg/sql/sem/tree/datum.go#L5070)):*
> ```Go
> }
> 
> // wrapAsZeroOid wraps and oid as zero oid value
> ```

nit: replace with
```
// wrapAsZeroOid wraps ZeroOidValue with a custom Oid.
```
___
*[pkg/sql/sem/tree/parse_string.go, line 66 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/71873#-MnbixGd89xEaO9kB35l:-MnbixGd89xEaO9kB35m:b-g36kad) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e2252a4a470cba43191d59700755d2bb20809cf/pkg/sql/sem/tree/parse_string.go#L66)):*
> ```Go
> 	case types.OidFamily:
> 		if s == ZeroDOidValue {
> 			d = NewDString(s)
> ```

i don't think this is right. it should be `d = NewDOid(0)`


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
also, after all the commits are combined, the commit message can be simplified to
```
    sql: fix casting of zero OID

    Release note (sql change): To align with PostgreSQL, casting an OID type
    with a value of `0` to a regtype, regproc, regclass, or regnamespace now
    will convert the value to the string `-`. The reverse behavior is
    implemented too, so a `-` will become `0` if casted to OID.
```

the reason is we don't want all the intermediate commits to pollute our git history

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71873#-:-MnblQYg6IEnrZjc81bK:b-qwbth5)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @lpessoa)



<!-- Sent from Reviewable.io -->

lpessoa commented:
@rafiss thanks for the feedback and tips for the condensed tests. I found another edge case when performing 
`select '-'::regclass`; on Postgres it returns `-` but on CDB it returns `0`.

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71873)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @lpessoa)

___
*[pkg/sql/sem/tree/datum.go, line 5044 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/71873#-MnbjnNd9FS2418i6-uI:-Mnfn5A8EC-QNJxUnslm:b1n9miu) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e2252a4a470cba43191d59700755d2bb20809cf/pkg/sql/sem/tree/datum.go#L5044)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: replace with
```
// ZeroOidValue postgres represents the 0 oid value as '-'.
```
</blockquote></details>

oops, i meant
```
// ZeroOidValue represents the 0 oid value as '-', which matches the Postgres representation.
```


<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71873)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @lpessoa)

___
*[pkg/sql/logictest/testdata/logic_test/cast, line 342 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/71873#-MnbimHn4yLIj6JHvem4:-MngCTgI2DAhymSx_8-h:b-dc7m3c) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e2252a4a470cba43191d59700755d2bb20809cf/pkg/sql/logictest/testdata/logic_test/cast#L342)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

all of these tests can be condensed a bit. you can just add these tests:

```
query IOTTTTTT
SELECT i, i::oid, i::oid::text,
  i::oid::regproc, i::oid::regprocedure, i::oid::regnamespace, i::oid::regclass, i::oid::regtype
FROM (VALUES (0), (1)) v(i)
----
0  0  0  -  -  -  -  -
1  1  1  1  1  1  1  1

query TOOOOO
SELECT i, i::regproc::oid, i::regprocedure::oid, i::regnamespace::oid, i::regclass::oid, i::regtype::oid
FROM (VALUES ('-')) v(i)
----
-  0  0  0  0  0

statement error invalid input syntax for type oid: "-"
SELECT '-'::oid
```
</blockquote></details>

and a few more:
```
query IOTOOOOOO
SELECT i, i::oid, i::oid::text,
  i::oid::regproc, i::oid::regprocedure, i::oid::regnamespace, i::oid::regclass, i::oid::regtype, i::oid::regrole
FROM (VALUES (0), (1)) v(i)
----
0  0  0  -  -  -  -  -  -
1  1  1  1  1  1  1  1  1

query TOOOOOOOOOOOO
SELECT i, i::regproc::oid, i::regprocedure::oid, i::regnamespace::oid, i::regtype::oid, i::regclass::oid, i::regrole::oid,
  i::regproc, i::regprocedure, i::regnamespace, i::regtype, i::regclass, i::regrole
FROM (VALUES ('-')) v(i)
----
-  0  0  0  0  0  0  -  -  -  -  -  -

statement error invalid input syntax for type oid: "-"
SELECT i, i::oid FROM (VALUES ('-')) v(i)

statement error invalid input syntax for type oid: "-"
SELECT '-'::oid

query OOOOOOOOOOOO
SELECT '-'::regclass, '-'::regclass::oid,
  '-'::regproc, '-'::regproc::oid,
  '-'::regprocedure, '-'::regprocedure::oid,
  '-'::regnamespace, '-'::regnamespace::oid,
  '-'::regtype, '-'::regtype::oid,
  '-'::regrole, '-'::regrole::oid
----
-  0  -  0  -  0  -  0  -  0  -  0
```
___
*[pkg/sql/sem/tree/cast.go, line 2350 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/71873#-MngBwDw0jWhQVcO9GaZ:-MngBwDw0jWhQVcO9Ga_:b-ovlb1q) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e2252a4a470cba43191d59700755d2bb20809cf/pkg/sql/sem/tree/cast.go#L2350)):*
> ```Go
> 			if string(*v) == ZeroDOidValue {
> 				// zero oid value is represented by '-' in postgres.
> 				tmpOid := NewDOid(0)
> ```

instead of this, this code can just call `wrapAsZeroOid()`
___
*[pkg/sql/sem/tree/datum.go, line 5072 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/71873#-MngC0tYCVnNfaOBu41l:-MngC0tYCVnNfaOBu41m:bzmxhh) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e2252a4a470cba43191d59700755d2bb20809cf/pkg/sql/sem/tree/datum.go#L5072)):*
> ```Go
> // wrapAsZeroOid wraps and oid as zero oid value
> func wrapAsZeroOid(oid oid.Oid) Datum {
> 	return wrapWithOid(NewDString(ZeroDOidValue), oid)
> ```

instead of delegating i think we can get our desired behavior by changing this function like:
```
// wrapAsZeroOid wraps ZeroOidValue with a custom Oid.
func wrapAsZeroOid(t *types.T) Datum {
	dOid := NewDOid(0)
	dOid.semanticType = t
	if t.Oid() != oid.T_oid {
		dOid.name = ZeroOidValue
	}
	return dOid
}
```
___
*[pkg/sql/sem/tree/parse_string.go, line 65 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/71873#-MngCEabEK1b4H8sAt1M:-MngCEabEK1b4H8sAt1N:b-a4hg2j) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e2252a4a470cba43191d59700755d2bb20809cf/pkg/sql/sem/tree/parse_string.go#L65)):*
> ```Go
> 		d, err = ParseDJSON(s)
> 	case types.OidFamily:
> 		if s == ZeroDOidValue {
> ```

this also should be:
```
        if t.Oid() != oid.T_oid && s == ZeroOidValue {
            d = wrapAsZeroOid(t)
         } else {
           // ...
```


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
i just edited some of the comments in the code.

CI failed due to a timezone test that was fixed in 2ff004d794, so this lgtm and will be merged. thanks for all your work on this!

bors r+
